{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d0880246",
   "metadata": {},
   "source": [
    "Multiple Myeloma data challenge\n",
    "============================\n",
    "Technical test for Data Scientist (DS) position in the Computational Biology team\n",
    "\n",
    "Katarzyna Wreczycka 17.10.2022\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Clinical context\n",
    "-------------------\n",
    "Multiple Myeloma (MM) is a type of bone marrow cancer. Treatment for MM involves\n",
    "combinations of drugs over multiple cycles. There is huge heterogeneity in treatment\n",
    "response with some individuals being non-responders and some patients remaining well for\n",
    "some time before a relapse. A better characterization of patients who relapse early can\n",
    "influence the treatment options and combinations.\n",
    "\n",
    "In this test, we propose to develop a model for predicting the risk of dying or relapsing of\n",
    "newly diagnosed multiple myeloma patients from baseline clinical and expression data.\n",
    "\n",
    "\n",
    "Data\n",
    "--------------------------\n",
    "The data for this test are extracted from an old Synapse Dream Challenge\n",
    "(https://www.synapse.org/#!Synapse:syn6187098/wiki/401884) .\n",
    "It consists of clinical data, gene expression data and follow-up for newly diagnosed Multiple\n",
    "Myeloma patients extracted from the MMRF CoMMpass IA9 study. In the data, newly\n",
    "diagnosed MM patients are classified as High Risk (HR) when they relapse or die before 18\n",
    "months.\n",
    "To access the data, you first need to create an account and download the following files:\n",
    "- Expression data:\n",
    "MMRF_CoMMpass_IA9_E74GTF_Salmon_entrezID_TPM_hg19.csv\n",
    "(https://www.synapse.org/#!Synapse:syn10573789)\n",
    "[notice that the first column gives Entrez IDs for genes]\n",
    "- Clinical data and labels:\n",
    "sc3_Training_ClinAnnotations.csv\n",
    "(https://www.synapse.org/#!Synapse:syn9926878)\n",
    "- Explanation of the clinical and label annotations:\n",
    "Harmonized_Clinical_Dictionary\n",
    "(https://www.synapse.org/#!Synapse:syn9744732)\n",
    "\n",
    "\n",
    "\n",
    "Goal\n",
    "--------------------------------------\n",
    "The purpose of this technical test is to develop a model for predicting the risk of fast dying or\n",
    "relapsing of newly diagnosed multiple myeloma patients (using the High Risk label\n",
    "HR_FLAG).The evaluation will mostly rely on the way you approach the problem: pre-analysis,\n",
    "preprocessing strategy, choice of modelization and coding skills.\n",
    "The code should be developed so that the model can be applied to an external validation\n",
    "dataset. You will send your code (Notebook or script) along with a small report to interpret\n",
    "the model and put it in MM context (the use of the literature is clearly welcome).\n",
    "Your model can be developed in Python or R with a small README to explain how to apply\n",
    "it to external data.\n",
    "\n",
    "You can use external knowledge/data to develop the model. Please add all the requirements\n",
    "for libraries that should be installed to make it run.\n",
    "\n",
    "If not used to survival analysis, the candidate can consider a simplified version in which it\n",
    "can assumed that no censored patients will be present in the external validation dataset.\n",
    "\n",
    "(Obvious) suggestion: OS and PFS related variables are also labels and not features:\n",
    "HR_FLAG is defined as OS or PFS < 18 months (taking into account censoring).\n",
    "\n",
    "--------------------------\n",
    "-----------------------------\n",
    "\n",
    "OS = overall survival\n",
    "\n",
    "PFS = progression-free survival = The length of time during and after the treatment of a disease, such as cancer, that a patient lives with the disease but it does not get worse. In a clinical trial, measuring the progression-free survival is one way to see how well a new treatment works.\n",
    "\n",
    "Censoring = a type of missing data problem unique to survival analysis. This happens when you track the sample/subject through the end of the study and the event never occurs. This could also happen due to the sample/subject dropping out of the study for reasons other than death, or some other loss to followup. The sample is censored in that you only know that the individual survived up to the loss to followup, but you don’t know anything about survival after tha\n",
    "\n",
    "\n",
    "high risk = defined as disease progression or death prior to 18 months from diagnosis\n",
    "\n",
    "\n",
    "in the publication (https://www.nature.com/articles/s41375-020-0742-z) they say something about \n",
    "https://blog.datadive.net/selecting-good-features-part-i-univariate-selection/\n",
    "\"The top-performing model implemented a wisdom of the crowd approach, utilizing clinical features and published myeloma signatures that summarize the expression of gene sets. The second-place “Stanford University Go” (SUGO) model instead included individual genes as features, utilizing a univariate-based feature selection approach to identify genes to include in their model\"\n",
    "\n",
    "(REE-laps) The return of a disease or the signs and symptoms of a disease after a period of improvement\n",
    "\n",
    "\n",
    "expression data is TPM (transcripts per million).\n",
    "\n",
    "\n",
    "The International Staging System (ISS) for multiple myeloma defines 3 subgroups with differing overall survival: Stage I- 62 months. Stage II- 44 months. Stage III- 29 months.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0212d605",
   "metadata": {},
   "source": [
    "Results\n",
    "-----------------------\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e89abaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load libraries\n",
    "\n",
    "suppressPackageStartupMessages(library(tidyverse));\n",
    "suppressPackageStartupMessages(library(data.table));\n",
    "suppressPackageStartupMessages(library(org.Hs.eg.db));\n",
    "suppressPackageStartupMessages(library(stats));\n",
    "\n",
    "# visualization\n",
    "suppressPackageStartupMessages(library(ggfortify)); \n",
    "suppressPackageStartupMessages(library(ggplot2));\n",
    "\n",
    "\n",
    "suppressPackageStartupMessages(library(glmnet));\n",
    "suppressPackageStartupMessages(library(survival));\n",
    "suppressPackageStartupMessages(library(ggplot2));\n",
    "suppressPackageStartupMessages(library(maxstat))\n",
    "\n",
    "\n",
    "library(stats)\n",
    "library(ggfortify); \n",
    "library(ggplot2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae455568",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data\n",
    "clinical_data = read.csv(\"./Harmonized_Clinical_Dictionary.csv\", sep=\",\", header=TRUE,fill = TRUE, stringsAsFactors = FALSE)\n",
    "expression_data = read.csv(\"./MMRF_CoMMpass_IA9_E74GTF_Salmon_entrezID_TPM_hg19.csv\", sep=\",\", header=TRUE,stringsAsFactors = FALSE,row.names=1)\n",
    "annot_data = read.csv(\"./sc3_Training_ClinAnnotations.csv\", sep=\",\", header=TRUE, stringsAsFactors = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebe3d909",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert the row names to entrez ids\n",
    "#library(\"AnnotationDbi\")\n",
    "#library(\"org.Hs.eg.db\")\n",
    "#geneSymbols <- mapIds(org.Hs.eg.db, keys=rownames(expression_data), \n",
    "#                      column=\"SYMBOL\", keytype=\"ENTREZID\", \n",
    "#                      multiVals=\"first\")\n",
    "#head(geneSymbols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f27e9aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "print(clinical_data[which(clinical_data$names==\"Study\"),]$description)\n",
    "table(annot_data$Study)\n",
    "\n",
    "\n",
    "print(\"PatientType\")\n",
    "table(annot_data$PatientType)\n",
    "\n",
    "\n",
    "print(clinical_data[which(clinical_data$names==\"D_Gender\"),]$description)\n",
    "table(annot_data$D_Gender)\n",
    "\n",
    "print(clinical_data[which(clinical_data$names==\"D_Age\"),]$description)\n",
    "summary(annot_data$D_Age)\n",
    "\n",
    "\n",
    "print(\"HR_FLAG\")\n",
    "table(annot_data$HR_FLAG)\n",
    "\n",
    "\n",
    "print(clinical_data[which(clinical_data$names==\"D_ISS\"),]$description)\n",
    "table(annot_data$D_ISS)\n",
    "\n",
    "print(clinical_data[which(clinical_data$names==\"D_OS_FLAG\"),]$description)\n",
    "table(annot_data$D_OS_FLAG)\n",
    "\n",
    "\n",
    "print(clinical_data[which(clinical_data$names==\"D_OS\"),]$description)\n",
    "summary(annot_data$D_OS)\n",
    "\n",
    "print(clinical_data[which(clinical_data$names==\"D_OS_FLAG\"),]$description)\n",
    "table(annot_data$D_OS_FLAG)\n",
    "\n",
    "\n",
    "print(clinical_data[which(clinical_data$names==\"D_PFS\"),]$description)\n",
    "summary(annot_data$D_PFS)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae48553a",
   "metadata": {},
   "outputs": [],
   "source": [
    "##############################\n",
    "\n",
    "print(head(annot_data$D_ISS))\n",
    "a = as.factor(as.character(annot_data$D_ISS))\n",
    "levels(a) <- list(\"I\"=\"1\",\"II\"=\"2\",\"III\"=\"3\")\n",
    "annot_data$D_ISS = a\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed2e3a0f",
   "metadata": {},
   "source": [
    "Data preprocessing\n",
    "---------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c12bd95",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=10)\n",
    "\n",
    "\n",
    "# we organized the crowd sourced Multiple Myeloma DREAM Challenge, \n",
    "# focusing on predicting high risk, defined as disease progression \n",
    "# or death prior to 18 months from diagnosis. \n",
    "\n",
    "\n",
    "\n",
    "#----------------- Preprocess the input data\n",
    "\n",
    "edata = expression_data\n",
    "\n",
    "\n",
    "# make sure that I all samples that I am going to analyse \n",
    "# have also labels\n",
    "colnames(edata) = gsub(\"_1_BM\",\"\",colnames(edata))\n",
    "intersect_labels = intersect(colnames(edata),annot_data$Patient)\n",
    "edata_m = edata[, match(intersect_labels, colnames(edata), nomatch = 0)]\n",
    "annot_data_m = annot_data[match(intersect_labels, annot_data$Patient, nomatch = 0), ]\n",
    "#sum(colnames(edata_m)==annot_data_m$Patient) == length(colnames(edata_m))\n",
    "\n",
    "# TODO: color-code it by HR FLAF, and labels 45 degrees\n",
    "bp <- boxplot(edata_m[,1:50],outline=FALSE,col=\"cornflowerblue\", xaxt = \"n\", yaxt = \"n\")\n",
    "tick <- seq_along(bp$names)\n",
    "axis(1, at = tick, labels = FALSE)\n",
    "text(tick, par(\"usr\")[3] - 0.3, bp$names, srt = 45, xpd = TRUE)\n",
    "\n",
    "#library(reshape2)\n",
    "#mt = melt(edata_m[,1:50])\n",
    "#ggplot(mt, aes(x = variable, y = value)) +\n",
    "#  geom_boxplot()\n",
    "#print(head(edata_m[,1:50]))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e20faa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#----------------- Filtering data and scaling\n",
    "\n",
    "# convert a value to a Standard Score (\"z-score\")\n",
    "\n",
    "# transform the expression data to Z scores\n",
    "#edata_sc <- t(scale(t(edata_m)))\n",
    "\n",
    "# Since taking a log seems to work to tame the extreme values, \n",
    "# we do that below and also add 1 pseudo-count to be able to deal with 0 values\n",
    "gexp=log10(edata_m+1)\n",
    "\n",
    "# transpose the data set\n",
    "tgexp <- t(gexp)\n",
    "\n",
    "library(caret)\n",
    "# remove near zero variation for the columns at least 85% of the values are the same\n",
    "# this function creates the filter but doesn't apply it yet\n",
    "nzv=preProcess(tgexp, method=\"nzv\", uniqueCut = 15)\n",
    "\n",
    "# apply the filter using \"predict\" function\n",
    "# return the filtered dataset and assign it to nzv_tgexp variable\n",
    "nzv_tgexp=predict(nzv,tgexp)\n",
    "\n",
    "#anyNA(nzv_tgexp) # check if there are NA values\n",
    "\n",
    "\n",
    "#gene.quant <- apply(beer.exprs.sample, 1, quantile, probs = 0.75)\n",
    "#dat.filt <- beer.exprs.sample[gene.quant > log2(100),]\n",
    "#gene.iqr <- apply(dat.filt, 1, IQR)\n",
    "#dat.filt <- as.matrix(dat.filt[gene.iqr > 0.5,])\n",
    "#dat.filt <- t(dat.filt)\n",
    "#dat.filt <- data.frame(dat.filt)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c357c9c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check if there are NA values in the expression matrix\n",
    "if(anyNA(nzv_tgexp)) print(\"There are NAs in the expression matrix\")\n",
    "\n",
    "# show \n",
    "nzv_gexp = t(nzv_tgexp)\n",
    "boxplot(nzv_gexp[,1:50],outline=FALSE,col=\"cornflowerblue\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21ebd2e4",
   "metadata": {},
   "source": [
    "Clustering\n",
    "---------------------\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b4c726c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let’s select the top 1000 most variable genes among the samples.\n",
    "options(repr.plot.width=20, repr.plot.height=10)\n",
    "\n",
    "\n",
    "library(pheatmap)\n",
    "\n",
    "#compute the variance of each gene across samples\n",
    "V <- apply(nzv_gexp, 1, var)\n",
    "#sort the results by variance in decreasing order \n",
    "#and select the top 100 genes \n",
    "selectedGenes <- names(V[order(V, decreasing = T)][1:1000])\n",
    "\n",
    "\n",
    "#Now we can quickly produce a heatmap where samples and genes are clustered.\n",
    "#library(pheatmap)\n",
    "#pheatmap(nzv_gexp[selectedGenes,], scale = 'row', show_rownames = FALSE)\n",
    "\n",
    "\n",
    "annot_sel_col <- c(\"D_Age\", \n",
    "                   \"D_Gender\", \n",
    "                   \"D_OS\",\n",
    "                   \"D_PFS\", \n",
    "                   \"D_ISS\", \n",
    "                   \"HR_FLAG\"\n",
    "                  )\n",
    "colData <- annot_data_m[which(colnames(annot_data_m) %in% annot_sel_col)]\n",
    "rownames(colData) = colnames(nzv_gexp)\n",
    "\n",
    "#print(head(nzv_gexp[selectedGenes,]))\n",
    "#print(head(colData))\n",
    "\n",
    "library(dichromat)\n",
    "colramp = colorRampPalette(c(3,\"white\",2))(9)\n",
    "\n",
    "pheatmap(nzv_gexp[selectedGenes,], \n",
    "         scale = 'row', \n",
    "         show_rownames = FALSE, \n",
    "         show_colnames = FALSE, \n",
    "         annotation_col = colData,\n",
    "         col=colramp,\n",
    "         main=\"All patients\")\n",
    "\n",
    "\n",
    "# Plot newly classified patients that relapsed or died before 18 months (including censored)\n",
    "pheatmap(nzv_gexp[selectedGenes,which(colData$D_PFS<547.501)], \n",
    "         scale = 'row', \n",
    "         show_rownames = FALSE, \n",
    "         show_colnames = FALSE, \n",
    "         annotation_col = colData[which(colData$D_PFS<547.501),],\n",
    "         col=colramp,\n",
    "         main=\"High-risk patients\")\n",
    "\n",
    "pheatmap(nzv_gexp[selectedGenes,which(colData$D_PFS>=547.501)], \n",
    "         scale = 'row', \n",
    "         show_rownames = FALSE, \n",
    "         show_colnames = FALSE, \n",
    "         annotation_col = colData[which(colData$D_PFS>=547.501),],\n",
    "         col=colramp,\n",
    "        main=\"Low-risk patients\")\n",
    "\n",
    "\n",
    "# https://rviews.rstudio.com/2017/09/25/survival-analysis-with-r/\n",
    "# https://www.biostars.org/p/344233/\n",
    "# https://cran.r-project.org/web/packages/glmnet/vignettes/Coxnet.pdf\n",
    "# https://compgenomr.github.io/book/gene-expression-analysis-using-high-throughput-sequencing-technologies.html\n",
    "# https://medium.com/towards-data-science/five-tips-on-survival-analysis-for-a-data-scientist-ba9fd97cbb2d\n",
    "# https://www.randomforestsrc.org/articles/getstarted.html\n",
    "# https://bioconnector.github.io/workshops/r-survival.html#tcga\n",
    "# https://rviews.rstudio.com/2017/09/25/survival-analysis-with-r/\n",
    "# https://www.biostars.org/p/344233/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "382eb744",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=7, repr.plot.height=7)\n",
    "\n",
    "\n",
    "# PCA\n",
    "# Let’s make a PCA plot to see the clustering of replicates as a scatter plot in two dimensions \n",
    "#transpose the matrix \n",
    "#M <- t(nzv_gexp[selectedGenes,])\n",
    "#M <- t(nzv_gexp)\n",
    "\n",
    "#nzv_gexp_centered2 = t(nzv_gexp) - colMeans(nzv_gexp)\n",
    "\n",
    "# transform the counts to log2 scale \n",
    "#M <- log2(M + 1)\n",
    "#compute PCA \n",
    "pcaResults <- prcomp(t(nzv_gexp), center = TRUE)\n",
    "\n",
    "#plot PCA results making use of ggplot2's autoplot function\n",
    "#ggfortify is needed to let ggplot2 know about PCA data structure.\n",
    "\n",
    "annot_data_m$D_ISS = as.factor(annot_data_m$D_ISS)\n",
    "\n",
    "#autoplot(pcaResults)\n",
    "\n",
    "autoplot(pcaResults, data = annot_data_m, colour = 'D_Gender')\n",
    "autoplot(pcaResults, data = annot_data_m, colour = 'D_ISS')\n",
    "autoplot(pcaResults, data = annot_data_m, colour = 'D_Age')\n",
    "autoplot(pcaResults, data = annot_data_m, colour = 'HR_FLAG')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31745236",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Modify annotation table\n",
    "y = annot_data_m\n",
    "y$time = y$D_PFS\n",
    "y$status = y$D_PFS_FLAG\n",
    "y$high.risk <- unlist(apply(y[, c(\"status\", \"time\")], 1,\n",
    "                                function(row) ifelse((row[1] == 1) && (row[2] < 18*(365.25/12)), \"high\", \"low\")))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc584f42",
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(pcaResults$)\n",
    "#dim(pcaResults$x)\n",
    "#print(dim(annot_data_m))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57e76371",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "eb61d63b",
   "metadata": {},
   "source": [
    "Correct gene expression data using the ComBat algorithm\n",
    "------------------------------------------------\n",
    "\n",
    "Looking at the results of the hierarchical custering and PCA above, I'll try to remove batch effects and other unwanted variation.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f71bb56",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "9428a023",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd743857",
   "metadata": {},
   "outputs": [],
   "source": [
    "#edata = expression_data[rowMeans(expression_data) > 100,]\n",
    "# to be ona scale thats easier to work with\n",
    "#edata = log2(edata + 1)\n",
    "# if we dont center the data, if we dont remove rowmeans of the center, the column means of data\n",
    "# then the first singual value will always be the mean level\n",
    "# since that will always explain the most variation in genomics experiment\n",
    "# we want to see variation between sample or between genes\n",
    "#edata_centered = edata - rowMeans(edata) \n",
    "#svd1 = svd(edata_centered)\n",
    "#names(svd1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b7a117e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# try KNN, Support Vector Machines and Ensemble Methods like Random Forests or Gradient Boosting"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0018bb9",
   "metadata": {},
   "source": [
    "Baseline models: age, ISS, and age + ISS\n",
    "---------------------------------------------------------\n",
    "\n",
    "Baseline models were Cox proportional hazards models where PFS was the response variable and age, ISS, or both age and ISS were explanatory variables. Continuous risk predictions scores of each sample were thresholded to create high risk classifications. The threshold of a given model was calculated by generating sample level  predictions in the training data. The threshold was defined as the cutpoint that maximized the logrank statistic comparing above and below the threshold to their true high-risk status. Continuous predictions for validation data were computed by applying the above trained Cox proportional hazards model to them. These values were dichotomized into high and low risk according to whether the prediction was above or below, respectively, the threshold.\n",
    "\n",
    "The baseline predictors were Cox proportional hazard models where progression free survival was modeled on age or ISS or age and ISS together. As comparators these models represent a reasonable lower bound on expected performance of high-risk classifiers. \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cea6490a",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=8, repr.plot.height=8)\n",
    "\n",
    "# Here is a survival analysis based on the one of the top models, \n",
    "# I am not sure though that it works well\n",
    "\n",
    "coxph.model = function(y, clinical.covariates){\n",
    "\n",
    "    surv.formula <- as.formula(paste(\"Surv(time, status)\", \"~\",\n",
    "                                     paste(clinical.covariates, collapse = \" + \"), sep = \" \"))\n",
    "    cph.fit <- coxph(surv.formula, data = y)\n",
    "    \n",
    "    pred <- unname(predict(cph.fit, newdata = y, type = \"risk\"))\n",
    "    y$pred <- pred\n",
    "\n",
    "    ## Calculate a threshold for the Cox proportional hazards model                                                                                                                     \n",
    "    mt.formula <- as.formula(paste(\"Surv(time, status)\", \"~\", \"pred\", sep = \" \"))\n",
    "    mt <- maxstat.test(mt.formula, data=data.frame(y), smethod=\"LogRank\", pmethod=\"none\")\n",
    "    print(mt$estimate)\n",
    "    \n",
    "    return(\n",
    "        list(y = y, \n",
    "             fitted.model = cph.fit, \n",
    "             threshold = as.numeric(mt$estimate))\n",
    "        )\n",
    "    \n",
    "}\n",
    "\n",
    "\n",
    "############## age\n",
    "                            \n",
    "res = coxph.model(y, \"D_Age\")\n",
    "threshold <- res$threshold\n",
    "y$pred = res$y$pred\n",
    "y$inferred.high.risk <- y$pred > threshold\n",
    "\n",
    "                            \n",
    "fisher.test(y$high.risk, y$inferred.high.risk)\n",
    "g <- ggplot(data = y, aes(x = high.risk, y = pred))\n",
    "g <- g + geom_boxplot()  +   geom_jitter()  + ggtitle(\"Age\") + ylab(\"Prediction\") + xlab(\"Defined risk\")             \n",
    "g     \n",
    "    \n",
    "                            \n",
    "cox_fit <- survfit(res$fitted.model)\n",
    "#plot(cox_fit, main = \"cph model\", xlab=\"Days\")\n",
    "autoplot(cox_fit) + ggtitle(\"Cox proportional hazard model: Age\") + xlab(\"Days\") + ylab(\"Survival\")\n",
    "                            \n",
    "                            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80271839",
   "metadata": {},
   "outputs": [],
   "source": [
    "############## ISS\n",
    "                            \n",
    "res = coxph.model(y, \"D_ISS\")\n",
    "threshold <- res$threshold\n",
    "y$pred = res$y$pred\n",
    "y$inferred.high.risk <- y$pred > threshold\n",
    "\n",
    "                            \n",
    "fisher.test(y$high.risk, y$inferred.high.risk)\n",
    "g <- ggplot(data = y, aes(x = high.risk, y = pred))\n",
    "g <- g + geom_boxplot()  +   geom_jitter()  + ggtitle(\"ISS\") + ylab(\"Prediction\") + xlab(\"Defined risk\")             \n",
    "g    \n",
    "\n",
    "cox_fit <- survfit(res$fitted.model)\n",
    "#plot(cox_fit, main = \"cph model\", xlab=\"Days\")\n",
    "autoplot(cox_fit) + ggtitle(\"Cox proportional hazard model: ISS\") + xlab(\"Days\") + ylab(\"Survival\")\n",
    "     \n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "334bf926",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=8, repr.plot.height=8)\n",
    "\n",
    "############## age + ISS  \n",
    "                            \n",
    "                            \n",
    "res = coxph.model(y, \"D_Age + D_ISS\")\n",
    "threshold <- res$threshold\n",
    "y$pred = res$y$pred\n",
    "y$inferred.high.risk <- y$pred > threshold\n",
    "\n",
    "                            \n",
    "fisher.test(y$high.risk, y$inferred.high.risk)\n",
    "g <- ggplot(data = y, aes(x = high.risk, y = pred))\n",
    "g <- g + geom_boxplot()  +   geom_jitter()  + ggtitle(\"D_Age + D_ISS\") + ylab(\"Prediction\") + xlab(\"Defined risk\")             \n",
    "g \n",
    "\n",
    "\n",
    "cox_fit <- survfit(res$fitted.model)\n",
    "#plot(cox_fit, main = \"cph model\", xlab=\"Days\")\n",
    "autoplot(cox_fit) + ggtitle(\"Cox proportional hazard model: Age+ISS\") + xlab(\"Days\") + ylab(\"Survival\")\n",
    "     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f18ebc4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(survAUC)\n",
    "\n",
    "head(y)\n",
    "#res = coxph.model(y, \"D_Age + D_ISS\")\n",
    "#threshold <- res$threshold\n",
    "#y$pred = res$y$pred\n",
    "#y$inferred.high.risk <- y$pred > threshold\n",
    "\n",
    "# 5-fold cross-validation repeated 5 times\n",
    "set.seed(123)\n",
    "kfold=5\n",
    "n=nrow(y)\n",
    "n2sample = floor(n / 5)\n",
    "\n",
    "indx.train = sample(n, n2sample)\n",
    "indx.test = c(1:n)[-indx.train]\n",
    "ytest = y[indx.test,]\n",
    "ytrain=y[indx.train,]\n",
    "#length(indx.train)\n",
    "#length(indx.test)\n",
    "#dim(ytrain)\n",
    "#dim(ytest)\n",
    "\n",
    "\n",
    "intauc_list = c()\n",
    "for (i in 1:100){\n",
    "    \n",
    "   \n",
    "    clinical.covariates = \"D_Age + D_ISS\"\n",
    "    surv.formula <- as.formula(paste(\"Surv(time, status)\", \"~\",\n",
    "                                     paste(clinical.covariates, collapse = \" + \"), sep = \" \"))\n",
    "    train.fit <- coxph(surv.formula, data = ytrain, method=\"breslow\",x=TRUE, y=TRUE)\n",
    "    \n",
    "    lp <- predict(train.fit)\n",
    "    lpnew <- predict(train.fit, newdata=ytest)\n",
    "    Surv.rsp <- survival::Surv(ytrain$time, ytrain$status)\n",
    "    Surv.rsp.new <- survival::Surv(ytest$time, ytest$status)\n",
    "    times <- seq(from=365, to=730, by=7) # from 12 months to 2 years every week\n",
    "    \n",
    "    AUC_CD <- AUC.cd(Surv.rsp, Surv.rsp.new, lp, lpnew, times)\n",
    "    \n",
    "    intauc=IntAUC(AUC_CD$auc, AUC_CD$times, runif(length(times),0,1), \n",
    "                  median(times), auc.type=\"cumulative\")\n",
    "    intauc_list = c(intauc_list, intauc)\n",
    "\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87b2f772",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(intauc_list)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "835ba811",
   "metadata": {},
   "outputs": [],
   "source": [
    "train.fit  <- survival::coxph(survival::Surv(futime, fustat) ~ age,\n",
    "                              x=TRUE, y=TRUE, method=\"breslow\", data=TR)\n",
    "\n",
    "lp <- predict(train.fit)\n",
    "lpnew <- predict(train.fit, newdata=TE)\n",
    "Surv.rsp <- survival::Surv(TR$futime, TR$fustat)\n",
    "Surv.rsp.new <- survival::Surv(TE$futime, TE$fustat)\n",
    "times <- seq(10, 1000, 10)                  \n",
    "times\n",
    "\n",
    "AUC_CD <- AUC.cd(Surv.rsp, Surv.rsp.new, lp, lpnew, times)\n",
    "\n",
    "AUC_CD\n",
    "\n",
    "runif(length(times),0,1)\n",
    "IntAUC(AUC_CD$auc, AUC_CD$times, runif(length(times),0,1), \n",
    "       median(times), auc.type=\"cumulative\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63b14377",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(survAUC)\n",
    "\n",
    "data(cancer,package=\"survival\")\n",
    "TR <- ovarian[1:16,]\n",
    "TE <- ovarian[17:26,]\n",
    "train.fit  <- survival::coxph(survival::Surv(futime, fustat) ~ age,\n",
    "                              x=TRUE, y=TRUE, method=\"breslow\", data=TR)\n",
    "\n",
    "lp <- predict(train.fit)\n",
    "lpnew <- predict(train.fit, newdata=TE)\n",
    "Surv.rsp <- survival::Surv(TR$futime, TR$fustat)\n",
    "Surv.rsp.new <- survival::Surv(TE$futime, TE$fustat)\n",
    "times <- seq(10, 1000, 10)                  \n",
    "times\n",
    "\n",
    "AUC_CD <- AUC.cd(Surv.rsp, Surv.rsp.new, lp, lpnew, times)\n",
    "\n",
    "AUC_CD\n",
    "\n",
    "runif(length(times),0,1)\n",
    "IntAUC(AUC_CD$auc, AUC_CD$times, runif(length(times),0,1), \n",
    "       median(times), auc.type=\"cumulative\")\n",
    "\n",
    "\n",
    "summary(TE$futime)\n",
    "\n",
    "# head(ovarian)\n",
    "#\tfutime\tfustat\tage\tresid.ds\trx\tecog.ps#\n",
    "#\t<dbl>\t<dbl>\t<dbl>\t<dbl>\t<dbl>\t<dbl>\n",
    "#1\t59\t1\t72.3315\t2\t1\t1\n",
    "#2\t115\t1\t74.4932\t2\t1\t1\n",
    "#3\t156\t1\t66.4658\t2\t1\t2\n",
    "#4\t421\t0\t53.3644\t2\t2\t1\n",
    "#5\t431\t1\t50.3397\t2\t1\t1\n",
    "#6\t448\t0\t56.4301\t1\t1\t2\n",
    "\n",
    "#The training dataset was split using repeated k-fold cross-validation \n",
    "#(specifically, 5 folds repeated 5 times) and used to calculate the average iAUC \n",
    "#over folds and repeats. This is the reference out-of-sample performance of the model.\n",
    "#Then the observed values of one of the variables were permuted and the predictions\n",
    "#and average iAUC were recalculated. The difference between the reference performance\n",
    "#and the new performance after permuting the variable defined the importance of that \n",
    "#variable. This procedure is then repeated for all variables in the model.\n",
    "\n",
    "\n",
    "#myModel = cv.glmnet(data.matrix(modelData), \n",
    "#                    modelData$ACTION,\n",
    "#                    family = \"binomial\",\n",
    "#                    type.measure = \"auc\",\n",
    "#                    nfolds = 5,\n",
    "#                    alpha = 1)\n",
    "# https://stats.stackexchange.com/questions/97777/variablity-in-cv-glmnet-results\n",
    "### cycle for doing 100 cross validations\n",
    "### and take the average of the mean error curves\n",
    "### initialize vector for final data.frame with Mean Standard Errors\n",
    "#MSEs <- NULL\n",
    "#for (i in 1:100){\n",
    "#                 cv <- cv.glmnet(y, x, alpha=alpha, nfolds=k)  \n",
    "#                 MSEs <- cbind(MSEs, cv$cvm)\n",
    "#             }\n",
    "#  rownames(MSEs) <- cv$lambda\n",
    "#  lambda.min <- as.numeric(names(which.min(rowMeans(MSEs))))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#Five-fold cross-validation was simulated 100 times to estimate prediction model performance. \n",
    "#The final model was comprised of one regularized logistic regression model trained on a\n",
    "#balanced training dataset, \n",
    "#classifying between high-risk and low-risk Multiple Myeloma.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# https://neptune.ai/blog/balanced-accuracy\n",
    "# balanced accuracy = (sensitivty + specificity) / 2\n",
    "\n",
    "# how to plot ROC curve\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2016f95d",
   "metadata": {},
   "source": [
    "Find a gene list fitting the cox model\n",
    "------------------------------------------------------\n",
    "\n",
    "\"The principal components or penalized Cox regression method is employed to select a gene list for fitting the Cox proportional harzards model. It provides an assessment of whether the association of expression data to survival data is statistically significant. It also lets the user evaluate whether the expression data provides more accurate predictions than that provided by standard clinical covariates.\"\n",
    "\n",
    "https://stats.stackexchange.com/questions/411804/how-to-choose-the-best-combination-of-covariates-in-cox-multiple-regression\n",
    "Identifying small sets of genes whose expression is related to outcome can readily be done with LASSO. It would select a subset of genes that together are strongly related to outcome while penalizing their regression coefficients to minimize overfitting. The R glmnet package provides tools for analyzing Cox models, including cross-validation to choose the optimum penalty factor (based on partial likelihood deviance in Cox models) and thus the number of genes chosen. LASSO can't avoid the dependence of the gene set chosen on the particular sample at hand, and you still should consider incorporating standard clinical variables into your model, but it provides a principled and well accepted approach to this type of problem\n",
    "\n",
    "elastic net might be alos ok\n",
    "https://accio.github.io/machine-learning/2018/01/04/glmnet.html\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44925357",
   "metadata": {},
   "outputs": [],
   "source": [
    "y_annot = annot_data_m[,c(\"D_PFS\",\"D_OS_FLAG\")]\n",
    "colnames(y_annot) = c(\"time\", \"status\")\n",
    "rownames(y_annot) = annot_data_m$Patient\n",
    "\n",
    "x_geneexpr = t(nzv_gexp)\n",
    "colnames(x_geneexpr) = paste0(\"gene\",colnames(x_geneexpr))\n",
    "\n",
    "x_geneexpr.age.iss = cbind(annot_data_m[,c(\"D_Age\",\"D_ISS\")], x_geneexpr)\n",
    "\n",
    "#head(annot_df)\n",
    "#x_geneexpr.age.iss[1:3, 1:3]\n",
    "as.matrix(x_geneexpr.age.iss)[1:3, 1:3]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e7759f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# remove patients that ISS is NA\n",
    "x_geneexpr.age.iss.nona = x_geneexpr.age.iss[-which(is.na(x_geneexpr.age.iss$D_ISS)),]\n",
    "y_annot.nona = y_annot[-which(is.na(x_geneexpr.age.iss$D_ISS)),]\n",
    "\n",
    "#a = apply(x_geneexpr.age.iss.nona, 1, function(x) which(is.na(x)))\n",
    "#b = apply(x_geneexpr.age.iss.nona, 2, function(x) which(is.na(x)))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f04fe385",
   "metadata": {},
   "outputs": [],
   "source": [
    "#head(x_geneexpr.age.iss.nona)\n",
    "#a\n",
    "#b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c84d066f",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(glmnet)\n",
    "library(survival)\n",
    "\n",
    "fit.elasticnet.noclinvar <- glmnet(as.matrix(x_geneexpr), as.matrix(y_annot), family = \"cox\", alpha=0.5) #TODO:alpha=1 by default\n",
    "fit.lasso.noclinvar <- glmnet(as.matrix(x_geneexpr), as.matrix(y_annot), family = \"cox\", alpha = 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46c023b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "fit.elasticnet.age.iss <- glmnet(x_geneexpr.age.iss.nona, as.matrix(y_annot.nona), family = \"cox\", alpha=0.5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0af462c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "fit.lasso.age.iss <- glmnet(x_geneexpr.age.iss.nona, as.matrix(y_annot.nona), family = \"cox\", alpha = 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a84bbee",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot(fit.elasticnet.noclinvar, xvar = \"lambda\", label = TRUE)\n",
    "\n",
    "\n",
    "head(coef(fit.elasticnet.noclinvar, s = 0.05))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39f6ef89",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot(fit.lasso.noclinvar, label=TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "deb40193",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot(fit.elasticnet.age.iss, label=TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad103f31",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot(fit.lasso.age.iss, label=TRUE)\n",
    "\n",
    "\n",
    "#library(biomaRt)\n",
    "#mart <- useMart(dataset=\"hsapiens_gene_ensembl\",biomart='ensembl')\n",
    "#gene_names = biomaRt::select(mart, \n",
    "#                             keys=c(\"148\"),\n",
    "#                             columns=c('hgnc_symbol'), \n",
    "#                             keytype='entrezgene_id')\n",
    "#gene_names\n",
    "# ADRA1A https://pubmed.ncbi.nlm.nih.gov/30061158/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "493836de",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8c8e83a",
   "metadata": {},
   "outputs": [],
   "source": [
    "set.seed(123)\n",
    "cvfit.lasso.noclinvar <- cv.glmnet(x_geneexpr, as.matrix(y_annot), family = \"cox\", type.measure = \"C\", alpha = 1)\n",
    "cvfit.elasticnet.noclinvar <- cv.glmnet(x_geneexpr, as.matrix(y_annot), family = \"cox\", type.measure = \"C\", alpha = 0.5)\n",
    "\n",
    "cvfit.lasso.age.iss <- cv.glmnet(x_geneexpr.age.iss.nona, as.matrix(y_annot.nona), \n",
    "                                family = \"cox\", type.measure = \"C\", alpha = 1)\n",
    "cvfit.elasticnet.age.iss <- cv.glmnet(x_geneexpr.age.iss.nona, as.matrix(y_annot.nona), \n",
    "                                 family = \"cox\", type.measure = \"C\", alpha = 0.5)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f97c33b",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "print(\"cvfit.lasso.noclinvar\")\n",
    "plot(cvfit.lasso.noclinvar )\n",
    "print(\"cvfit.elasticnet.noclinvar\")\n",
    "plot(cvfit.elasticnet.noclinvar)\n",
    "\n",
    "print(\"cvfit.lasso.age.iss\")\n",
    "plot(cvfit.lasso.age.iss)\n",
    "\n",
    "print(\"cvfit.elasticnet.age.iss\")\n",
    "plot(cvfit.elasticnet.age.iss)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "406481ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "cvfit.lasso.noclinvar$lambda.min\n",
    "cvfit.elasticnet.noclinvar$lambda.min\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af86463e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# http://stats.stackexchange.com/questions/124288/r-glmnet-cross-validated-auc\n",
    "        # cv.glmnet fits a whole sequence of models, and will report the auc for all of them. \n",
    "        # The max of the cvm sequence is the best model's auc.\n",
    "        # cvm = mean cross-validated error - a vector of length length(lambda)\n",
    "#        auc.val = max(model$models[[3]]$cvm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84146d53",
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://github.com/bswhite/Celgene-Multiple-Myeloma-Challenge-Baseline-Models/blob/master/publishedClassifiers/lasso.weighted.ensemble/scratch.R\n",
    "\n",
    "folds <- cvFolds(n=nrow(all.training.uams),K=10)\n",
    "# Function to minimize to identify the best-fit alpha value\n",
    "train <- function(par,numeric.matrix,folds,classes,mc) {\n",
    "  library(glmnet)\n",
    "  library(doParallel)\n",
    "  registerDoParallel(cl = mc)\n",
    "  model <- cv.glmnet(y=classes,x=numeric.matrix,family = 'cox',foldid = folds$which,parallel = T)\n",
    "  model$cvm[model$lambda == model$lambda.1se]\n",
    "}\n",
    "mc <- makeCluster(detectCores())\n",
    "classes <- as(Surv(time=as.numeric(all.training.uams$PFStimeMonths),event=all.training.uams$Progression),'matrix')\n",
    "numeric.matrix=as.matrix(all.training.uams[,-c(1,2)])\n",
    "optimized <- optim(par=0,train,method='Brent',lower=0,upper=1,numeric.matrix=numeric.matrix,folds=folds,classes=classes,mc)\n",
    "model <- cv.glmnet(x=numeric.matrix,y=classes,family='cox',alpha=optimized$par,foldid = folds$which,parallel = T)\n",
    "stopCluster(mc)\n",
    "coef <- predict(model,newx=numeric.matrix,s=model$lambda.1se,type='coefficients')\n",
    "\n",
    "\n",
    "\n",
    "# https://github.com/bswhite/Celgene-Multiple-Myeloma-Challenge-Baseline-Models/blob/master/publishedClassifiers/common/classifier-base.R\n",
    "## Assume data is gene x samples.  This standardizes the genes so that each row has\n",
    "## mean 0 and variance 1.\n",
    "standardize.data <- function(data) {\n",
    "    means <- rowMeans(data)\n",
    "    sds <- apply(data,1,sd)\n",
    "    data <- ((data - means) / sds)\n",
    "    data\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "fit.and.threshold.coxph.model <- function(y, clinical.covariates) {\n",
    "\n",
    "    ## Fit a Cox proportional hazards model\n",
    "    surv.formula <- as.formula(paste(\"Surv(time, event)\", \"~\",\n",
    "                                     paste(clinical.covariates, collapse = \" + \"), sep = \" \"))\n",
    "    cph.fit <- coxph(surv.formula, data = y)\n",
    "\n",
    "    pred <- unname(predict(cph.fit, newdata = y, type = \"risk\"))\n",
    "    y$cph <- pred\n",
    "\n",
    "    ## Calculate a threshold for the Cox proportional hazards model                                                                                                                     \n",
    "    suppressPackageStartupMessages(library(maxstat))\n",
    "    mt.formula <- as.formula(paste(\"Surv(time, event)\", \"~\", \"cph\", sep = \" \"))\n",
    "    mt <- maxstat.test(mt.formula, data=data.frame(y), smethod=\"LogRank\", pmethod=\"none\")\n",
    "\n",
    "    list(y = y, fitted.model = cph.fit, threshold = as.numeric(mt$estimate))\n",
    "\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa2223d6",
   "metadata": {},
   "source": [
    "Classification via logistic regression\n",
    "----------------------------------------\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed5d55a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Its a top performing method\n",
    "\n",
    "# The RNA-Seq data was first normalized using log transformation, \n",
    "# and both the microarray and RNA-Seq data were next standardized using z-scores for all genes.\n",
    "\n",
    "\n",
    "# 1) genes mapped to certain pathways and chromosomal abnormalities\n",
    "#    I. Chromosomal abnormalities [3]: deletion of chromosome 1p, \n",
    "#                                        gain of chromosome 1q, \n",
    "#                                        gain of chromosome 9, \n",
    "#                                        deletion of chromosome 13q, \n",
    "#                                        deletion of chromosome 17p, \n",
    "#                                        translocation t(4;14), \n",
    "#                                        trainslocation t(11;14), \n",
    "#                                        translocation t(14;16/14;20).\n",
    "#    II. DNA repair pathways [5]: non-homologous end-joining pathway\n",
    "#                                 homologous recombination pathway, \n",
    "#                                 Fanconi anemia pathway, \n",
    "#                                 nucleotide excision repair pathway, \n",
    "#                                 mismatch repair pathway, \n",
    "#                                 base excision repair pathway.\n",
    "#    III. Other pathways [3]: cell cycle pathway, \n",
    "#                              p53 signalling pathway, \n",
    "#                              NF-kB signalling pathway, \n",
    "#                              Ras-ERK pathway.\n",
    "#    IV. Genes targeted by Multiple Myeloma treatments [6]: Bortezomib, Thalidomide -> AURKA, IGF1,  AURKA (3), IGF1R (5), and FGFR3\n",
    "#    V. Mutations associated with high-risk Multiple Myeloma [3].\n",
    "#    VI. Other gene expression profiles obtained from literature: EMC92 [7], UAMS70 [8], \n",
    "#                       DNA repair pathway score [4], -> BUB1, BUB1B|PAK6, RAD51, PLK1, BRCA1, CENPA, BARD1, AURKA, MAD2L1, CENPH, XRCC2 and CDC25C|FAM53C)\n",
    "#                       IFM group [9], -> TP53 \n",
    "#                       cell death network [10].\n",
    "# 2) gene assocaited with myeloma\n",
    "\n",
    "# Sum of gene expressions were used, and feature selection of the engineered features \n",
    "# was also performed to progressively discard uninformative features. \n",
    "# We also include clinical data comprising age and ISS as features.\n",
    "\n",
    "\n",
    "# Samples with missing data were discarded for the training dataset, \n",
    "# and only genes that were found across all training and validation datasets were retained. \n",
    "# To minimize the problem of class imbalance in training, random undersampling was performed to \n",
    "# ensure equal class distribution between high-risk and low-risk samples in the training dataset. \n",
    "# The machine learning algorithm chosen for the construction of the final prediction model was regularized \n",
    "# logistic regression, implemented using the glmnet package[11] in R. Five-fold cross-validation was simulated \n",
    "# 100 times to estimate prediction model performance. The final model was comprised of one regularized logistic \n",
    "# regression model trained on a balanced training dataset, classifying between high-risk and low-risk \n",
    "# Multiple Myeloma.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c1ba652",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create the final model to be saved\n",
    "    cv.glmn=cv.glmnet(x= data.list[[i]],\n",
    "                      y=y.list[[i]],\n",
    "                      family=\"binomial\",\n",
    "                      type.measure=\"auc\",alpha=alpha)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f76bd29f",
   "metadata": {},
   "source": [
    "A univariate approach to rank gene for feature selection in building expression based models\n",
    "------------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abcba3e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# differential gene expession\n",
    "# diff expr on gene expression doesnt make sense since I have only TPMs values\n",
    "# https://support.bioconductor.org/p/98820/\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69b0fd08",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Second Place Feature Selection Methods\n",
    "\n",
    "# a simple univariate approach to rank gene for feature selection in building expression based models.\n",
    "# In each of the four training datasets they computed each gene’s effect size, z, via the concordance\n",
    "# index between overall survival and the gene’s expression. These effect sizes were then combined across\n",
    "# training sets using Stouffer’s method with no weighting to yield a single meta-z per gene. They employed\n",
    "# this meta concordance index method under two expression normalization procedures with CDKN3 and PHF19 \n",
    "# coming out on top respectively and combined them with clinical features to create their model.\n",
    "# For more information on SUGO’s model please see their full description in Synapse\n",
    "# (https://www.synapse.org/#!Synapse:syn10380508/wiki/499377)\n",
    "\n",
    "\n",
    "# https://www.biostars.org/p/344233/\n",
    "# test each gene independently via Cox regression\n",
    "\n",
    "library(survival)\n",
    "library(RegParallel)\n",
    "\n",
    "#print(dim(coxdata))\n",
    "#exit()\n",
    "\n",
    "#res_featuresel <- RegParallel(\n",
    "   # data = coxdata,\n",
    "   # formula = 'Surv(time, status) ~ [*]',\n",
    "   # FUN = function(formula, data)\n",
    "   #   coxph(formula = formula,\n",
    "  #      data = data,\n",
    "  #      ties = 'breslow',\n",
    " #       singular.ok = TRUE),\n",
    " #   FUNtype = 'coxph',\n",
    " #   variables = colnames(coxdata)[3:ncol(coxdata)],\n",
    "#    blocksize = 2000,\n",
    "#    cores = 3,\n",
    "#    nestedParallel = FALSE,\n",
    "#    conflevel = 95)\n",
    "\n",
    "#saveRDS(res_featuresel,\"./res_featuresel.RDS\")\n",
    "#res_featuresel = readRDS(\"./res_featuresel.RDS\")\n",
    "\n",
    "#res <- res[!is.na(res$P),]\n",
    "#head(res_featuresel)\n",
    "\n",
    "#res_featuresel <- res_featuresel[order(res_featuresel$LogRank, decreasing = FALSE),]\n",
    "#final <- subset(res_featuresel, LogRank < 0.0001)\n",
    "#final\n",
    "#print(dim(coxdata))\n",
    "#print(dim(final))\n",
    "#head(final)\n",
    "#final_entrez_ids = gsub(\"gene\",\"\",final$Variable)\n",
    "\n",
    "#print(head(res_featuresel ))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b140f83",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d1af253",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(biomaRt)\n",
    "mart <- useMart(dataset=\"hsapiens_gene_ensembl\",biomart='ensembl')\n",
    "\n",
    "#hgnc_symbol_ids = mclapply(final_entrez_ids, function(x){\n",
    "#    biomaRt::select(mart, \n",
    "#                    keys=x,\n",
    "#                    columns=c('hgnc_symbol'), \n",
    "#                    keytype='entrezgene_id')\n",
    "#}, mc.cores=2)\n",
    "\n",
    "gene_names = biomaRt::select(mart, \n",
    "                             keys=final_entrez_ids,\n",
    "                             columns=c('hgnc_symbol'), \n",
    "                             keytype='entrezgene_id')\n",
    "res_featuresel = cbind(res_featuresel, \n",
    "                       gene_name = gene_names)\n",
    "head(res_featuresel)\n",
    "#entrezgene_ids = as.character(rownames(expression_data))\n",
    "#hgnc_symbol_ids = gene_names$hgnc_symbol\n",
    "#rownames(expression_data) = hgnc_symbol_ids\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14bb0120",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d605ba1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#head(res_featuresel)\n",
    "#head(gene_names[,1])\n",
    "#length(gene_names[,1])\n",
    "#nrow(res_featuresel)\n",
    "#head(res_featuresel)\n",
    "#dim(res_featuresel)\n",
    "#write.csv(gene_names,\"./final_gene_names.csv\")\n",
    "#write.table(gene_names, \"./final_gene_names.txt\", append = FALSE, sep = \" \", dec = \".\",\n",
    "#            row.names = FALSE, col.names = FALSE, quote = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1abed58e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#surv.formula <- as.formula(paste(\"Surv(time, status)\", \"~\",\"1\", sep = \" \"))\n",
    "#print(surv.formula)\n",
    "#cph.fit <- coxph(surv.formula, data = coxdata)\n",
    "#head(cph.fit)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f8d56c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#head(annot_data_m)\n",
    "#annot_df = annot_data_m[,c(\"D_PFS\",\"D_OS_FLAG\", \"D_Age\", \"D_ISS\")]\n",
    "#colnames(annot_df) = c(\"time\", \"status\", \"age\", \"ISS\")\n",
    "#coxdata <- cbind(annot_df, t(edata_m))\n",
    "#colnames(coxdata) = as.character(colnames(coxdata))\n",
    "#head(coxdata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfcd79e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# \"EZH2\",\"MCM4\",\"TYMS\",\"AURKB\",\"CHEK1\",\"CHEK1\",\"ZWINT\",\"CCNA2\",\"BIRC5\"\n",
    "# \"PHF19\", \"MMSET\", \"D_AGE\", \"D_ISS\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9287356",
   "metadata": {},
   "source": [
    "Survival modelling using Random Forest and Gradient Boosting\n",
    "------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14488b6f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "1ac74857",
   "metadata": {},
   "source": [
    "Compare models\n",
    "----------------------------\n",
    "\n",
    "https://seandavi.github.io/TargetOsteoAnalysis/articles/multivariate_survival.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43b92938",
   "metadata": {},
   "outputs": [],
   "source": [
    "#surv_all = survival::coxph(\n",
    "#  survival::Surv(time,status) ~ metastatic + CTA_chrX + CTA_all,\n",
    "#  data=ssgsea_survival)\n",
    "#anova(surv_all)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5063c8b6",
   "metadata": {},
   "source": [
    "Assessing model performance\n",
    "----------------------------------\n",
    "The integrated area under the curve (iAUC) and balanced accuracy curve (BAC). While the AUC is a widely accepted metric of prediction accuracy, it is sensitive to the specific time threshold used to differentiate high and low patient risk.(https://www.nature.com/articles/s41375-020-0742-z#Tab1). The myeloma research community has not yet reached a consensus on the time point that best separates patients into risk groups, though there is a general agreement that it lies between 1 and 2 years for progression-free survival (PFS). We, therefore, chose the iAUC “centered” on 18 months as a more robust primary metric. The iAUC range began 6 months prior to 18 months and continued to 6 months past with a sliding PFS threshold moving from12 to 24 months at weekly increments. iAUCs computed in each validation cohort were combined into a weighted average (wiAUC) with each cohort iAUC weighted by the square of the number of high-risk patients in order to ensure larger studies with more high-risk patients did not overwhelm smaller studies while still weighting them more heavily.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3621363e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c731357b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04ab2957",
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5f73ded",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
